### PR TITLE
bento unit tests: flush all hooks/renders after each test run

### PR DESCRIPTION
**Pull request by samouri**
**summary**
Addresses https://github.com/ampproject/amphtml/issues/36239.

The Preact util function `act(callback):Promise` will return a promise which resolves when all of the hooks queued up during the callback have completed running. This is exactly what we need to ensure all hooks finish running. Unfortunately, since the hooks need to be queued up during execution of the callback, we can't use `afterEach`, and instead need something more like `aroundEach` which doesn't exist in Mocha (https://github.com/mochajs/mocha/issues/1153).

In terms of a general pattern we have a few options I can see:
1. Monkey patch `it` to wrap its test in `act`
2. Create a new function, `bit` (bento it) which does the same (calls `it` and wraps the given test in act)
3. Have each test suite wrap all of its unit tests with `act`. This would get repetitive pretty fast.
4. Find (or create) some function within `preact` which would enable us to flush the queue in an afterEach

----

@jridgewell  / @rsimha / @rileyajones: any preferences between 1-4?

### CODE REVIEW

**Comment by rsimha** [comment]
My top preference is 4, since I think it generally makes sense for test suites to clean up after themselves. I'm not opposed to inserting it into the global `afterEach()`, and call the cleanup code when we detect that a bento test has just been run (is this possible?)

Re: options 1, 2, and 3, I fear that they will introduce a new learning curve for developers who are already familiar with the existing `describe`, `describes`, and `it`, and risk the introduction of bugs due to failure to use the new testing patterns.

Will defer to @rileyajones who may have other considerations in mind.

FWIW, this is what our global cleanup code looks like today:

https://github.com/ampproject/amphtml/blob/f23e77d1428a64c5039887fd05f3e73581f506cf/testing/init-tests.js#L127-L189

**Comment by samouri** [comment]
Discussing with @developit, looks like we start an `act` within the beforeEach, and wait to resolve the promise until the afterEach:

```js
let flush, flushed;
beforeEach(() => {
  flushed = act(() => {
    return new Promise(resolve => {
      flush = resolve;
    });
  });
});
afterEach(async () => {
  flush();
  await flushed;
});
```
---

I'll try to get this working in our global configs

File: extensions/amp-iframe/1.0/component.js
Comment (rileyajones) [inline extensions/amp-iframe/1.0/component.js:75]: Is this an intended revert or did you not rebase after merging #36241?

File: src/preact/test-utils.js
Comment (rileyajones) [inline src/preact/test-utils.js:1]: I'm not sure this is the correct place for this file to live, this seems like something that should go under `testing`

File: extensions/amp-iframe/1.0/component.js
Comment (samouri) [reply extensions/amp-iframe/1.0/component.js:75]: Intentionally removed it, since this PR fixes the root cause of the issue.

File: src/preact/test-utils.js
Comment (jridgewell) [reply src/preact/test-utils.js:1]: Agreed.

File: src/preact/test-utils.js
Comment (jridgewell) [inline src/preact/test-utils.js:11]: Weakset?

Comment (samouri) [reply src/preact/test-utils.js:11]: Good idea. Cleaner than adding a key to all of the funcs. Done

File: src/preact/test-utils.js
Comment (samouri) [reply src/preact/test-utils.js:1]: Good thinking! Moved to `testing/preact`.

File: testing/init-tests.js
Comment (rsimha) [inline testing/init-tests.js:173]: Methinks this should be the first line of this function and not the last line. This way, everything related to test execution will be guaranteed to finish before we do things like restore sandboxes, delete page elements, reset global state, etc.

File: testing/preact.js
Comment (jridgewell) [inline testing/preact.js:28]: This is the one that should have been a weakset.

File: testing/preact.js
Comment (jridgewell) [inline testing/preact.js]: This one should actually a boolean, not a weakset. `process` is a module constant in preact, and we need to continually rerun it. But, it may be schedule multiple times before finally running, and we only want to run once.

```suggestion
/**
 * Ensure render is debounced
 */
const debounceRendered = true;

/**
 * @param {() => void} process
 */
function flushableRender(process) {
  flushRender = process;
  debounceRendered = false;
  Promise.resolve().then(() => {
    if (debounceRendered) {
      return;
    }
    debounceRendered = true;
    flushRender = null;
    return process();
  });
```

**Review by jridgewell — CHANGES_REQUESTED** [review]

File: testing/init-tests.js
Comment (samouri) [reply testing/init-tests.js]: Good idea! Done

File: testing/preact.js
Comment (developit) [reply testing/preact.js]: Yes - sorry, this was the reason for the `while (pending)` in my Gist that you asked about (just saw the comment now):
https://gist.github.com/developit/3d7a69a2e819c8d11888f3a310a98f8f#gistcomment-3921021

Any time the commit/flush callback from `debounceRendering` is invoked, all of the enqueued components are rendered. If a component enqueues additional updates during that render (while the queue is draining, but during/after rendering of that component), it will be enqueued in a second render pass that gets deferred by calling `options.debounceRendering`.

Here's an annotated version of my gist suggestion that better illustrates:

```js
// promise defined when any renders are enqueued:
// (resolves when the queue is flushed, undefined when idle)
let pending;

options.debounceRendering = f => {
  // the render queue now contains work (from setState, etc)
  pending = Promise.resolve().then(() => {
    // flush is synchronous, so we have to reset the indicator here to avoid
    // clobbering a new `pending` set by a nested debounceRendering call:
    pending = undefined;
    // flush all *currently* enqueued rendering:
    f();
  });
};

export async function flush() {
  // invoke any pending effects first, since they may enqueue renders (but cannot enqueue effects):
  await flushRaf();
  // if any renders were enqueued...
  while (pending) {
    await pending;  // ...wait for them to finish (just a Promise tick)
    await flushRaf();  // ...then invoke any effects enqueued by the render
  }
}
```

File: testing/preact.js
Comment (jridgewell) [inline testing/preact.js:28]: Nit: This is `n^2`. A simple for-of and truncate will be a bit faster.

File: testing/preact.js
Comment (jridgewell) [inline testing/preact.js:58]: There's a time-order bug if we await here, but luckily we don't need to await a sync function.

If we awaited here, then a render queued by an effect will be processed before we flush in this loop. That's this await takes 2 ticks (even for a sync return value), and the `flushableRender`'s pendingRender will itself fire on the 2nd tick. If the second render queued an effect, then we won't notice to recall `flushRaf`.
```suggestion
   flushRaf();
```

**Review by jridgewell — APPROVED** [review]

**Pull request merged**
